// @tanstack/query-core@5.0.0-alpha.88/build/modern/index.js
import { CancelledError } from "./retryer";
import { QueryCache } from "./queryCache";
import { QueryClient } from "./queryClient";
import { QueryObserver } from "./queryObserver";
import { QueriesObserver } from "./queriesObserver";
import { InfiniteQueryObserver } from "./infiniteQueryObserver";
import { MutationCache } from "./mutationCache";
import { MutationObserver } from "./mutationObserver";
import { notifyManager } from "./notifyManager";
import { focusManager } from "./focusManager";
import { onlineManager } from "./onlineManager";
import {
  hashKey,
  replaceEqualDeep,
  isServer,
  matchQuery,
  keepPreviousData
} from "./utils";
import { isCancelledError } from "./retryer";
import {
  dehydrate,
  hydrate,
  defaultShouldDehydrateQuery,
  defaultShouldDehydrateMutation
} from "./hydration";
export * from "./types";
export {
  CancelledError,
  InfiniteQueryObserver,
  MutationCache,
  MutationObserver,
  QueriesObserver,
  QueryCache,
  QueryClient,
  QueryObserver,
  defaultShouldDehydrateMutation,
  defaultShouldDehydrateQuery,
  dehydrate,
  focusManager,
  hashKey,
  hydrate,
  isCancelledError,
  isServer,
  keepPreviousData,
  matchQuery,
  notifyManager,
  onlineManager,
  replaceEqualDeep
};
